src/metabase/models/field_usage.clj

```diff
@@ -0,0 +1,100 @@
+(ns metabase.models.field-usage
+  (:require
+   [metabase.lib.core :as lib]
+   [metabase.lib.dispatch :as lib.dispatch]
+   [metabase.lib.hierarchy :as lib.hierarchy]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.util :as lib.util]
+   [metabase.lib.util.match :as lib.util.match]
+   [metabase.models.interface :as mi]
+   [metabase.query-processor.middleware.fetch-source-query :as fetch-source-query]
+   [metabase.util.malli :as mu]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]
+   [toucan2.tools.disallow :as t2.disallow]))
+
+(doto :model/FieldUsage
+  (derive :metabase/model)
+  (derive ::t2.disallow/update)
+  (derive :hook/created-at-timestamped?))
+
+(methodical/defmethod t2/table-name :model/FieldUsage [_model] :field_usage)
+
+(t2/deftransforms :model/FieldUsage
+  {:used_in                mi/transform-keyword
+   :aggregation_function   mi/transform-keyword
+   :breakout_temporal_unit mi/transform-keyword
+   :breakout_binning       mi/transform-json
+   :filter_op              mi/transform-keyword
+   :filter_args            mi/transform-json})
+
+(defn- filter->field-usage
+  [query stage filter-clause]
+  (let [filter-parts (lib/filter-parts query stage filter-clause)
+        field-id     (-> filter-parts :column :id)]
+    (when (int? field-id)
+      {:field_id    field-id
+       :used_in     :filter
+       :filter_op   (-> filter-parts :operator :short)
+       :filter_args (:args filter-parts)})))
+
+(defn- aggregation->field-usage
+  [query stage aggregation-clause]
+  (let [aggregation-column (lib/aggregation-column query stage aggregation-clause)
+        field-id           (:id aggregation-column)]
+    (when (int? field-id)
+      {:field_id             field-id
+       :used_in              :aggregation
+       :aggregation_function (first aggregation-clause)})))
+
+(defn- breakout->field-usage
+  [query stage breakout-clause]
+  (let [breakout-column (lib/breakout-column query stage breakout-clause)
+        field-id        (:id breakout-column)]
+    (when (int? field-id)
+      {:field_id               field-id
+       :used_in                :breakout
+       :breakout_temporal_unit (lib/raw-temporal-bucket breakout-clause)
+       :breakout_binning       (-> (lib/binning breakout-clause)
+                                   (select-keys [:strategy :bin-width :num-bins])
+                                   not-empty)})))
+
+(defn- expression->field-usage
+  [expresison-clause]
+  (when-let [field-ids (seq (lib.util.match/match expresison-clause [:field _ (id :guard int?)] id))]
+    (for [field-id field-ids]
+      {:field_id field-id
+       :used_in  :expression})))
+
+(defmulti ^:private pquery->field-usages-method
```

Comment:
Does this need to be a multimethod? From what I can see, for every call site of `pquery->field-usages-method` we know the dispatch type from the context, so there doesn't seem to be a need for any polymorphism here.

e.g. on [line 87](https://github.com/metabase/metabase/pull/39671/files#diff-8f297ee24e62bb71f2aed630a98593bee3f2a3e353ae8db0ab2c0edf08a03eeaR87) we already know that `x` is going to be an `mbql/join`:
```clojure
    (flatten (keep #(pquery->field-usages-method query stage-number %) (:joins x)))))
```

The only thing that unites the three implementations is that it takes _something_ and returns a sequence of field usages. But this could be three separate functions, each with more appropriate names and argument lists.

Reply:
I wrote a version that did not use multimethod at first. But I took Cam's advice and tried to follow how `returned-columns-method` is defined. 

It's a bitter sweet experience since it's a bit to grasp at first. But overall I find dealing with recursive using multimethod makes the implementation of each multimethods simpler.

Reply:
> overall I find dealing with recursive using multimethod makes the implementation of each multimethods simpler.

I needed to check this out for myself, because I thought I was missing something. I don't think I am.

Here's an approach without multimethods: 3 functions and 23 LOC.

```clojure
(declare pmbql->field-usages)

(defn- join->field-usages
  [query join]
  (let [join-query (fetch-source-query/resolve-source-cards (assoc query :stages (get join :stages)))]
    ;; treat the source query as a :mbql/query
    (pmbql->field-usages join-query)))

(defn- stage->field-usages
  [query stage-number]
  (concat
    (keep #(filter->field-usage query stage-number %) (lib/filters query stage-number))
    (keep #(aggregation->field-usage query stage-number %) (lib/aggregations query stage-number))
    (keep #(breakout->field-usage query stage-number %) (lib/breakouts query stage-number))
    (flatten (keep expression->field-usage (lib/expressions query stage-number)))
    (flatten (keep #(join->field-usages query %) (:joins x)))))

(mu/defn pmbql->field-usages
  "Given a pmbql query, returns field usages from filter, breakout, aggregation, expression of a query.
  Walk all stages and joins.
  Expects all the source cards were resolved"
  [query :- ::lib.schema/query]
  (mapcat #(stage->field-usages query %) (range (lib/stage-count query))))
```

vs the original: 4 functions (counting multimethod implementations) and 48 LOC

```clojure
(defmulti ^:private pmbql->field-usages-method
  "Impl for [[pmbql->field-usages]]."
  {:arglists '([query stage-number x])}
  (fn [_query _stage-number x] (lib.dispatch/dispatch-value x))
  :hierarchy lib.hierarchy/hierarchy)

(defmethod pmbql->field-usages-method :mbql/query
  [query _stage-number x]
  (flatten (for [stage (range (lib/stage-count x))]
                (pmbql->field-usages-method query stage (lib.util/query-stage query stage)))))

(defmethod pmbql->field-usages-method :metabase.lib.stage/stage
  [query stage-number x]
  (concat
    (keep #(filter->field-usage query stage-number %) (lib/filters query stage-number))
    (keep #(aggregation->field-usage query stage-number %) (lib/aggregations query stage-number))
    (keep #(breakout->field-usage query stage-number %) (lib/breakouts query stage-number))
    (flatten (keep expression->field-usage (lib/expressions query stage-number)))
    (flatten (keep #(pmbql->field-usages-method query stage-number %) (:joins x)))))

(defmethod pmbql->field-usages-method :mbql/join
  [query _stage-number x]
  (let [join-query (fetch-source-query/resolve-source-cards (assoc query :stages (get x :stages)))]
    ;; treat the source query as a :mbql/query
    (pmbql->field-usages-method join-query nil join-query)))

(mu/defn pmbql->field-usages
  "Given a pmbql query, returns field usages from filter, breakout, aggregation, expression of a query.
  Walk all stages and joins.
  Expects all the source cards were resolved"
  [pmbql :- ::lib.schema/query]
  (pmbql->field-usages-method pmbql nil pmbql))
```

Reply:
ok I'll bite. `defn` all the way in [ea4aff0](https://github.com/metabase/metabase/pull/39671/commits/ea4aff04d5d1a2d1c113db614a5b603841a0d45a)